hw/dv/verilator/cpp/scrambled_ecc32_mem_area.cc

```diff
@@ -0,0 +1,164 @@
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <sstream>
+
+#include "sv_scoped.h"
+
+// This is the maximum width of a nonce that's supported by the code in
+// prim_util_get_scramble_key_nonce.svh
+static const uint32_t kScrMaxNonceWidth = 320;
+static const uint32_t kScrMaxNonceWidthBytes = (kScrMaxNonceWidth + 7) / 8;
+
+// Functions to convert from integer address to/from a vector of bytes,
+// addr_width is given in bits
+static std::vector<uint8_t> AddrIntToBytes(uint32_t addr, uint32_t addr_width) {
+  uint32_t addr_width_bytes = (addr_width + 7) / 8;
+  std::vector<uint8_t> addr_bytes(addr_width_bytes);
+
+  for (int i = 0; i < addr_width_bytes; ++i) {
+    addr_bytes[i] = addr & 0xff;
+    addr >>= 8;
+  }
+
+  return addr_bytes;
+}
+
+static uint32_t AddrBytesToInt(const std::vector<uint8_t> &addr) {
+  uint32_t addr_out = 0;
+  int cur_shift = 0;
+
+  for (int i = 0; i < addr.size(); ++i) {
+    addr_out |= (addr[i] << cur_shift);
+    cur_shift += 8;
+  }
+
+  return addr_out;
+}
+
+extern "C" {
+int simutil_get_scramble_key(svBitVecVal *key);
+int simutil_get_scramble_nonce(svBitVecVal *nonce);
+}
+
+std::vector<uint8_t> ScrambledEcc32MemArea::GetScrambleKey() const {
+  SVScoped scoped(scr_scope_);
+  uint8_t key_minibuf[kPrinceWidthByte * 2];
+
+  if (!simutil_get_scramble_key((svBitVecVal *)key_minibuf)) {
+    std::ostringstream oss;
+    oss << "Could not read key at scope " << scr_scope_;
+    throw std::runtime_error(oss.str());
+  }
+
+  return std::vector<uint8_t>(&key_minibuf[0],
+                              &key_minibuf[kPrinceWidthByte * 2]);
+}
+
+std::vector<uint8_t> ScrambledEcc32MemArea::GetScrambleNonce() const {
+  assert(GetNonceWidthByte() <= kScrMaxNonceWidthBytes);
+
+  SVScoped scoped(scr_scope_);
+  uint8_t nonce_minibuf[kScrMaxNonceWidthBytes];
+
+  if (!simutil_get_scramble_nonce((svBitVecVal *)nonce_minibuf)) {
+    std::ostringstream oss;
+    oss << "Could not read nonce at scope " << scr_scope_;
+    throw std::runtime_error(oss.str());
+  }
+
+  return std::vector<uint8_t>(&nonce_minibuf[0],
+                              &nonce_minibuf[kScrMaxNonceWidthBytes]);
+}
+
+ScrambledEcc32MemArea::ScrambledEcc32MemArea(const std::string &scope,
+                                             uint32_t size, uint32_t width_32)
+    : Ecc32MemArea(
+          join_sv_scopes(scope,
+                         "u_prim_ram_1p_adv.u_mem.gen_generic.u_impl_generic"),
+          size, width_32),
+      scr_scope_(scope) {
+  // Calculate clog2(size) using integer arithmetic to avoid any edge cases
+  // introduced by float representation
+
+  // Begin with a minimum addr_width_ of 1
+  uint32_t calc_size_log2 = size >> 1;
+  addr_width_ = 1;
+
+  // Determine addr_width_ based on highest set bit of size
+  while (calc_size_log2 >>= 1) {
+    addr_width_++;
+  }
+
+  // For non power of two size addr_width_ must be incremented
+  if ((1 << addr_width_) < size) {
+    addr_width_++;
+  }
```

Comment:
Maybe it would be clearer to define the $clog2 function separately?
```c++
// Analogous to the vbits SystemVerilog function from prim_util_pkg.sv
static uint32_t vbits(uint32_t size) {
  uint32_t width = 0;
  while (size) {
    width++;
    size /= 2;
  }
  return width;
}
```
As a plus, this also matches an existing SV function, which should help people understand what it means.

This doesn't have the special behaviour that you've got at size == 0, but I think we should probably have an assertion at the start that `size > 0` anyway.

Reply:
Function now separate and named vbits for the sv utils counterpart. Note your implementation wasn't quite correct for powers of two, e.g. for size 8 you want 3 address bits, not 4. My comment was also incorrect as I'm not implementing clog2!